src/main/java/org/broadinstitute/hellbender/tools/walkers/bqsr/BaseRecalibratorUprooted.java

```diff
@@ -0,0 +1,564 @@
+                .and(MAPPING_QUALITY_NOT_ZERO)
+                .and(MAPPING_QUALITY_AVAILABLE)
+                .and(MAPPED)
+                .and(PRIMARY_ALIGNMENT)
+                .and(NOT_DUPLICATE)
+                .and(PASSES_VENDOR_QUALITY_CHECK);
+    }
+
+    private static SAMRecord consolidateCigar(SAMRecord read) {
+        // Always consolidate the cigar string into canonical form, collapsing zero-length / repeated cigar elements.
+        // Downstream code cannot necessarily handle non-consolidated cigar strings.
+        read.setCigar(AlignmentUtils.consolidateCigar(read.getCigar()));
+        return read;
+    }
+
+    private SAMRecord resetOriginalBaseQualities(SAMRecord read){
+        if (! useOriginalBaseQualities) {
+            return read;
+        }
+        return ReadUtils.resetOriginalBaseQualities(read);
+    }
+
+    private SAMRecord setDefaultBaseQualities(SAMRecord read) {
+        // if we are using default quals, check if we need them, and add if necessary.
+        // 1. we need if reads are lacking or have incomplete quality scores
+        // 2. we add if defaultBaseQualities has a positive value
+        if (defaultBaseQualities < 0) {
+            return read;
+        }
+        byte reads[] = read.getReadBases();
+        byte quals[] = read.getBaseQualities();
+        if (quals == null || quals.length < reads.length) {
+            byte new_quals[] = new byte[reads.length];
+            Arrays.fill(new_quals, defaultBaseQualities);
+            read.setBaseQualities(new_quals);
+        }
+        return read;
+    }
+
+    public RecalibrationTables getTables() {
+        return recalibrationEngine.getRecalibrationTables();
+    }
+
+
+    private ReadTransformer makeReadTransform(){
+        ReadTransformer f0 = BaseRecalibratorUprooted::consolidateCigar;
+
+        ReadTransformer f =
+                f0.andThen(this::setDefaultBaseQualities)
+                .andThen(this::resetOriginalBaseQualities)
+                .andThen(ReadClipper::hardClipAdaptorSequence)
+                .andThen(ReadClipper::hardClipSoftClippedBases);
+
+        return f;
+    }
+
+    @Override
+    public void apply( SAMRecord originalRead, ReferenceContext ref, FeatureContext featureContext ) {
+        throw new RuntimeException("apply Not implemented");
+    }
+
+    /**
+     * For each read at this locus get the various covariate values and increment that location in the map based on
+     * whether or not the base matches the reference at this particular location.
+     * The "features" list doesn't need to be complete but it must cover the read
+     */
+    public void apply( SAMRecord originalRead, final ReferenceContext ref, List<? extends Feature> features ) {
+        ReadTransformer transform = makeReadTransform();
+        final SAMRecord read = transform.apply(originalRead);
+
+        if( ReadUtils.isEmpty(read) ) { return; } // the whole read was inside the adaptor so skip it
```

Comment:
Please expand this comment. It's the hardClipAdaptorSequence that can result in a zero-length read. This happens when the read is exclusively part of the adaptor (whatever the case is if I didn't understand the code).